### DISCUSSION TITLE
Should we be reusing Client objects?

### DISCUSSION THREAD

**Discussion by tyteen4a03** [encode/httpx#1962] [Q&A]
In following to [this question](https://stackoverflow.com/questions/69916682/python-httpx-how-does-httpx-clients-connection-pooling-work) on SO:

Similar to the question [asked here](https://stackoverflow.com/questions/66534027/python-requests-httpx-repsonses-and-garbage-collection), we are seeing memory leaks in our HTTP client wrapper. We have a singleton that reuses an `httpx.Client()` and return the `Response` object back to the caller of our singleton. Our current suspicion is the `Client` not cleaning up properly.

Is this the correct pattern? How should we prevent this leak while making sure the entire `Response` object is still available to the downstream callers of our singleton?

**Comment by lovelydinosaur** [comment]
> We have a singleton that reuses an httpx.Client() and return the Response object back to the caller of our singleton.

That's fine yup. A singleton or a globally available `httpx.Client()` is perfectly sensible.

I can't really give you any guidance on the request body question without a specific example to work against, but worth noting that if you're uploading large requests you might want to stream them, rather than loading them into memory. Yes they'll stick around until the response is gc'ed.

**Reply by tyteen4a03** [reply]
I don't have one right now but does the [StackOverflow](https://stackoverflow.com/questions/66534027/python-requests-httpx-repsonses-and-garbage-collection) one help?

**Reply by lovelydinosaur** [reply]
So I can see some behaviour in the issue you've linked, but I can't see an issue.

* Does the response hold a reference to the request? *Yes*.
* If you're using a non-streaming request body will that continue to be in memory until your don't have a reference to either the request or response anymore? *Yes*.

That looks to me just like a necessary trade-off in "do we want the outgoing request to remain accessible on the response instance or not".
